File: Java/subTree.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {

    public static boolean isSame(TreeNode root, TreeNode subRoot){
        if(root==null && subRoot==null)
            return true;

        else if(root==null|| subRoot== null || root.val!=subRoot.val)
            return false;
        
        if(!isSame(root.left, subRoot.left))
            return false;
        
        if(!isSame(root.right, subRoot.right))
            return false;
        
        return true;        
    }

    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root==null)
            return false;
        
        if(root.val==subRoot.val){
            if(isSame(root,subRoot))
                return true;
        }

        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
}

File: Python/subTree.py
from typing import Optional
class TreeNode:
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right

class Solution:
    
    def isSame(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if root is None and subRoot is None:
            return True
        elif root is None or subRoot is None or root.val!=subRoot.val:
            return False
        
        if self.isSame(root.left, subRoot.left) is False:
            return False
        
        if self.isSame(root.right, subRoot.right) is False:
            return False
        
        return True

    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        if root is None:
            return False
        
        if root.val==subRoot.val:
            if self.isSame(root, subRoot) is True:
                return True
        
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)